---
title: "gene coverage analysis"
author: "Clalit Genomics"
date: "14/01/2022"
output: 
  html_document:
    theme: united
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
analysis_or_report<-'analysis'
bam_folder<-'/media/SSD/Bioinformatics/Projects/gene_coverage_analysis_2022/data/bam_analysis/input/'

project_dir<-'/media/SSD/Bioinformatics/Projects/gene_coverage_analysis_2022/'
knitr::opts_knit$set(root.dir='/media/SSD/Bioinformatics/Projects/gene_coverage_analysis_2022/')
knitr::opts_chunk$set(echo = F)
library(ProjectTemplate)
setwd(project_dir)
load.project()

#input_bams_df<-input_bams_df%>%mutate(group_name=stringr::str_match(group_name,'(agilent|idt|twist)')[,1])
```

```{r input_def}
# Setup Variables ####
#target_region<-'/media/SSD/Bioinformatics/Databases/refseq/refseq_hg19_curated_cds_20220403.bed.gz'
target_region<-'./data/target_intervals/HPD_v4_target_hg19.txt.bed'
#reference_fasta<-'/media/SSD/Bioinformatics/Databases/dragen/hg19_dragen.fa'
reference_fasta<-'/media/SSD/Bioinformatics/Databases/dragen/hg19_dragen.fa'
#gene_list_file<-'/media/SSD/Bioinformatics/Projects/gene_coverage_analysis_2022/data/gene_lists/mody_panel.csv'


```

```{r calculate_bam_coverage_with_bedtools}

main_output_folder<-glue('./output/bam_analysis_{Sys.Date()}')
if (!dir.exists(main_output_folder)){dir.create(main_output_folder)}

bam_analysis_df<-data.frame(input_bams_df,genome_file='',coverage_output='')

for (group_name in unique(input_bams_df$group_name)){
  message(glue('Analyzing {group_name}..'))
  #if (group_name!='idt'){next}
  group_output_folder<-glue('{main_output_folder}/{group_name}')
  if (!dir.exists(group_output_folder)){dir.create(group_output_folder)}
  for (sample_name in input_bams_df[input_bams_df$group_name==group_name,'sample_name']){
    message(glue('Analyzing {sample_name}..'))
    sample_output_folder<-glue('{group_output_folder}/{sample_name}')
    if (!dir.exists(sample_output_folder)){dir.create(sample_output_folder)}
    input_bam<-input_bams_df[input_bams_df$group_name==group_name & input_bams_df$sample_name==sample_name,'sample_path']
    # Prepare the data for coverage analsysis ####
    # prepare the genome file using the given bam file (required by bedtools)
    genome_file<-prepare_genome_file(input_bam,sample_output_folder)
    # prepare the input target region bed file
    # this file is generated using the ucsc table browser - generating a bed file with the CODING exons
    sorted_target_bed<-sort_input_bed_file(target_region,sample_output_folder,genome_file)
  
    # Run the bedtools coverage analysis #### 
    coverage_ouput<-run_bedtools_coverage_by_bam(input_bam,genome_file,sorted_target_bed,sample_output_folder)
    bam_analysis_df<-bam_analysis_df%>%rows_update(data.frame(group_name=group_name,
                                                              sample_name=sample_name,
                                                              genome_file=genome_file,
                                                              coverage_output=coverage_ouput),
                                                   by = c("group_name","sample_name"))
  }
}
bam_analysis_table<-glue('{main_output_folder}/bam_analysis_table.csv')
write.table(bam_analysis_df,file=bam_analysis_table,sep='\t',row.names = F)

```

```{r prep_per_group_coverage_ref}
bam_analysis_df_file<-glue('{main_output_folder}/bam_analysis_table.csv')
bam_analysis_df<-readr::read_delim(bam_analysis_df_file)
cov_per_exon_all_groups<-NULL
cov_per_transcript_all_groups<-NULL
for (i in 1:nrow(bam_analysis_df)){
#for (i in c(1,7)){
  bam_analysis_row<-slice(bam_analysis_df,i)
  group_name<-bam_analysis_row%>%pull(group_name)
  sample_name<-bam_analysis_row%>%pull(sample_name)
  message(glue('Analyzing {group_name}:{sample_name} ..'))
  bedtools_coverage_hist_output<-bam_analysis_row%>%pull(coverage_output)
  cov_data<-read_bed_coverage_hist_file(bedtools_coverage_hist_output)
  cov_per_exon<-summarize_coverage_hist_per_exon(cov_data)
  cov_per_exon_all_groups<-cov_per_exon_all_groups%>%bind_rows(data.frame(group_name=group_name,cov_per_exon))
  cov_per_transcript<-summarize_coverage_hist_per_transcript(cov_data)
  cov_per_transcript_all_groups<-cov_per_transcript_all_groups%>%bind_rows(data.frame(group_name=group_name,cov_per_transcript))
}

# Build ref per group ####
cov_per_transcript_per_group<-cov_per_transcript_all_groups%>%group_by(group_name,gene_name,transcript_name,chr,coverage_cat)%>%
  summarize(perc_at_coverage=mean(perc_at_cov_cat))
cov_per_exon_per_group<-cov_per_exon_all_groups%>%group_by(group_name,gene_name,transcript_name,exon_num,chr,coverage_cat)%>%
  summarize(perc_at_coverage=mean(perc_at_cov_cat))

write_group_coverage_ref<-function(selected_group_name){
  message(glue('Saving {selected_group_name} per-transcript and per-exon coverage reference'))
  write.table(cov_per_transcript_per_group%>%filter(group_name==selected_group_name),
              file=glue('{main_output_folder}/{selected_group_name}/{selected_group_name}_per_transcript_coverage_reference.tsv'),
              sep='\t',
              row.names = F)
  write.table(cov_per_exon_per_group%>%filter(group_name==selected_group_name),
              file=glue('{main_output_folder}/{selected_group_name}/{selected_group_name}_per_exon_coverage_reference.tsv'),
              sep='\t',
              row.names = F)
}
library(purrr)
map(unique(cov_per_transcript_per_group$group_name),write_group_coverage_ref)

```

go over the target file and output problematic regions for each input file

```{r analyze_coverage_by_target}
output_folder_to_parse<-'./output/bam_analysis_2023-08-20'
bam_analysis_def_table<-readr::read_delim(glue('{output_folder_to_parse}/bam_analysis_table.csv'))
threshs<-c(1,10,20)
all_cov_output<-NULL
per_target_cov_output<-NULL
for (i in 1:nrow(bam_analysis_def_table)){
  a_cov_output<-NULL
  pt_cov_output<-NULL
  group_name<-bam_analysis_def_table$group_name[i]
  sample_name<-bam_analysis_def_table$sample_name[i]
  bam_cov_file<-bam_analysis_def_table$coverage_output[i]
  bam_cov<-readr::read_delim(bam_cov_file,delim = '\t',col_names = c('chr','start','end','depth','num_bases','total_bases_in_target','percent_of_target'))
  all_cov<-bam_cov%>%filter(chr=='all')
  colnames(all_cov)<-c('all','depth','num_bases','total_bases_in_target','percent_of_entire_target','X1','X2')
  
  bam_cov<-bam_cov%>%filter(chr!='all')
  for (thresh in threshs){
    
    bases_col_name<-glue('bases_below_{thresh}x')
    pct_col_name<-glue('pct_below_{thresh}x')
    # All target analysis
    all_below_thresh<-all_cov%>%
      filter(depth<thresh)%>%
      group_by(all)%>%
      summarize(!!sym(bases_col_name):=sum(num_bases),
                !!sym(pct_col_name):=sum(percent_of_entire_target))
  
    # Per target analysis
    below_thresh<-bam_cov%>%
      filter(depth<thresh)%>%
      group_by(chr,start,end,total_bases_in_target)%>%
      summarize(!!sym(bases_col_name):=sum(num_bases),
                !!sym(pct_col_name):=sum(percent_of_target))
    if(is.null(pt_cov_output)){
      pt_cov_output<-below_thresh 
      a_cov_output<-all_below_thresh
    }else{
      pt_cov_output<-pt_cov_output%>%right_join(below_thresh)%>% replace(is.na(.), 0)
      a_cov_output<-a_cov_output%>%right_join(all_below_thresh)%>%replace(is.na(.), 0)
    }
  }
  a_cov_output<-a_cov_output%>%mutate(sample=sample_name,group=group_name)
  pt_cov_output<-pt_cov_output%>%mutate(sample=sample_name,group=group_name)
per_target_cov_output<-per_target_cov_output%>%bind_rows(pt_cov_output)
all_cov_output<-all_cov_output%>%bind_rows(a_cov_output)
}
all_cov_output

xlsx::write.xlsx(as.data.frame(all_cov_output),sheetName = 'All targets coverage',row.names = FALSE,file=glue('{output_folder_to_parse}/target_coverage_analysis.{Sys.Date()}.xls'))
xlsx::write.xlsx(as.data.frame(per_target_cov_output),sheetName = 'Per target coverage',row.names = FALSE,file=glue('{output_folder_to_parse}/target_coverage_analysis.{Sys.Date()}.xls'),append = TRUE)

```
