---
title: "ClinVar coverage analysis"
author: "Clalit Genomics"
date: "14/01/2022"
output: 
  html_document:
    theme: united
    toc: true
    toc_float: true
---

- BAM-based analysis:
    * Calculate the coverage for each clinvar variant in the input clinvar VCF - the output is saved in the output folder 
    * Parse the coverage output


```{r setup, include=FALSE}
analysis_setup_file<-'./config/analysis_setup/template.analysis_setup.txt'
project_dir<-'/media/SSD/Bioinformatics/Projects/gene_coverage_analysis_2022/'
knitr::opts_knit$set(root.dir=project_dir)
knitr::opts_chunk$set(echo = F)
library(ProjectTemplate)
setwd(project_dir)
load.project()
```

```{r intialize}
input_files<-readr::read_delim('./config/input_files.txt',delim='\t')
# analysis_setup
analysis_setup<-readr::read_delim(analysis_setup_file,delim='\t')
# analysis type
analysis_type<-analysis_setup%>%filter(param=='analysis_type')%>%pull(value)
# bam folder
bam_folder<-analysis_setup%>%filter(param=='bam_folder')%>%pull(value)
input_bams_df<-read_input_bam_files(bam_folder)
# clinvar files
clinvar_file_name<-analysis_setup%>%filter(param=='clinvar_file')%>%pull(value)
clinvar_file<-input_files%>%filter(type=='clinvar_snv',name==clinvar_file_name)%>%pull(file)
# target regions
target_regions_names<-analysis_setup%>%filter(param=='target_regions')%>%pull(value)%>%strsplit(',')%>%unlist()
target_regions_table<-input_files%>%inner_join(data.frame(name=target_regions_names))
target_regions<-target_regions_table%>%pull(file)
target_reference_regions<-target_regions_table%>%pull(ref)
per_target_analysis_req<-analysis_setup%>%filter(param=='per_target_analysis')%>%pull(value)%>%strsplit(',')%>%unlist()
# reference regions
reference_fasta<-input_files%>%filter(type=='reference_fasta')%>%inner_join(data.frame(ref=target_reference_regions))%>%pull(file)
reference_dict<-input_files%>%filter(type=='reference_dict')%>%inner_join(data.frame(ref=target_reference_regions))%>%pull(file)
# refseq transcript to gene name
refseq2gene<-parse_refseq_transcript_to_gene_name()

# prep output
main_output_prefix<-analysis_setup%>%filter(param=='main_output_folder_prefix')%>%pull(value)
main_output_folder<-glue('./output/{main_output_prefix}.clinvar_coverage_analysis.{Sys.Date()}')
bed_analysis_output_folder<-glue('{main_output_folder}/bed_analysis')
bam_analysis_output_folder<-glue('{main_output_folder}/bam_analysis')
if (!dir.exists(main_output_folder)){dir.create(main_output_folder)}
if (!dir.exists(bed_analysis_output_folder) & analysis_type%in%c('bed','both')){dir.create(bed_analysis_output_folder)}
if (!dir.exists(bam_analysis_output_folder) & analysis_type%in%c('bam','both')){dir.create(bam_analysis_output_folder)}

# copy analysis setup to the ouput folder
system(glue('cp {analysis_setup_file} {main_output_folder}'))

```

```{r prep_clinvar_file}
clinvar_bed<-fix_clinvar_ucsc_file(clinvar_file,main_output_folder,only_p_or_lp = T)
clinvar_file<-clinvar_bed
```

```{r clinvar_coverage_bed_based}
# Given a list of target bed files check how are the P/LP clinvar variants covered in the different regions
targets_vs_clinvar_df<-NULL
for (i in 1:nrow(target_regions_table)){
  tr<-target_regions_table%>%slice(i)%>%pull(name)
  tr_file<-target_regions_table%>%slice(i)%>%pull(file)
  message(glue('Analyzing {tr}'))
  clinvar_vs_target_file<-intersect_target_with_clinvar(clinvar_file = clinvar_file,
                                                        target_file = tr_file,
                                                        preprocess_input_clinvar = FALSE,
                                                        clinvar_vs_target_file = glue('{bed_analysis_output_folder}/{tr}_vs_clinvar.tsv'))
  # read the output file
  clinvar_vs_target<-read_clinvar_vs_target_file(clinvar_vs_target_file)
  targets_vs_clinvar_df<-targets_vs_clinvar_df%>%bind_rows(clinvar_vs_target%>%mutate(target=tr,target_file=tr_file,.before=1))
}
# Save all variants not covered in bed
targets_vs_clinvar_df_not_coverd<-targets_vs_clinvar_df%>%filter(num_o_overlap==0)
write.table(targets_vs_clinvar_df_not_coverd,file=glue('{bed_analysis_output_folder}/clinvar_not_covered_in_bed.{Sys.Date()}.txt'),sep='\t',row.names = F)

# Summarize coverage
targets_vs_clinvar_df<-targets_vs_clinvar_df%>%left_join(targets_vs_clinvar_df%>%group_by(var_type)%>%summarize(n=n()))
not_covered_summary<-targets_vs_clinvar_df%>%group_by(target,var_type,n)%>%
  summarize(num_not_covered=sum(num_o_overlap==0))%>%
  mutate(strip_name=glue('{var_type} ({n})'),
         rate=num_not_covered/n)
write.table(not_covered_summary,file=glue('{bed_analysis_output_folder}/clinvar_not_covered_in_bed_summary.{Sys.Date()}.txt'),sep='\t',row.names = F)


```

```{r clinvar_coverage_bam_based}
# This part takes in a clinvar file and input bam files and generates coverage output for each variant in the clinvar file

bam_analysis_df<-data.frame(input_bams_df)
coverage_df<-NULL

for (group_name in unique(input_bams_df$group_name)){
  message(glue('Analyzing {group_name}..'))
  #if (group_name!='idt'){next}
  group_output_folder<-glue('{bam_analysis_output_folder}/{group_name}')
  if (!dir.exists(group_output_folder)){dir.create(group_output_folder)}
  for (sample_name in input_bams_df[input_bams_df$group_name==group_name,'sample_name']){
    message(glue('Analyzing {sample_name}..'))
    sample_output_folder<-glue('{bam_analysis_output_folder}/{sample_name}')
    clinvar_cov_file<-glue('{sample_output_folder}/{sample_name}_clinvar_cov_metrics.csv')
    if (file.exists(clinvar_cov_file)){
      message(glue('{clinvar_cov_file} already exists, will skip running coverage calculation'))
    }else{
      if (!dir.exists(sample_output_folder)){dir.create(sample_output_folder)}
      input_bam<-input_bams_df[input_bams_df$group_name==group_name & input_bams_df$sample_name==sample_name,'sample_path']
      
      genome_file<-prepare_genome_file(input_bam,output_folder = sample_output_folder)
      
      calculate_clinvar_coverage(clinvar_bed = clinvar_bed,
                                 bam_file = input_bam,
                                 genome_file = genome_file,
                                 clinvar_cov_file = clinvar_cov_file)
      
    }
    
    clinvar_coverage_df<-read_clinvar_coverage_file(clinvar_cov_file)
    clinvar_coverage_metrics<-clinvar_coverage_df%>%group_by(var_type)%>%summarize(n=n(),
                                                         mean_depth=mean(depth),
                                                         x0_cov_vars=sum(depth==0),
                                                         below_x10_cov_vars=sum(depth<10),
                                                         below_x20_cov_vars=sum(depth<20))
    clinvar_coverage_metrics<-clinvar_coverage_metrics%>%mutate(group_name=group_name,
                                                                sample_name=sample_name,
                                                                .before=1)
    coverage_df<-coverage_df%>%bind_rows(clinvar_coverage_metrics)
  }
}

coverage_file_name<-glue('{bam_analysis_output_folder}/clinvar_coverage_analysis_results_{Sys.Date()}.tsv')
write.table(coverage_df,file=coverage_file_name,sep='\t',row.names = F)
```

```{r generate_clinvar_bam_coverage_report}
library(rmarkdown)
library(knitr)
render('/media/SSD/Bioinformatics/Projects/gene_coverage_analysis_2022/src/clinvar_coverage_analysis_report.Rmd', 
       output_dir = main_output_folder,
       output_file = glue('{main_output_folder}/{main_output_prefix}.clinvar_bam_coverage.{Sys.Date()}.html'),
       quiet = F,
       params=list(main_output_folder=main_output_folder,
                   main_output_prefix=main_output_prefix,
                   analysis_type=analysis_type))
```


