---
title: "BAM coverage analysis"
author: "Clalit Genomics"
date: "14/01/2022"
output: 
  html_document:
    theme: united
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
analysis_setup_file<-'./config/analysis_setup/template.analysis_setup.txt'
project_dir<-'/media/SSD/Bioinformatics/Projects/gene_coverage_analysis_2022/'
knitr::opts_knit$set(root.dir=project_dir)
knitr::opts_chunk$set(echo = F)
library(ProjectTemplate)
setwd(project_dir)
load.project()
```

```{r prep_analysis}
# prep output
main_output_prefix<-analysis_setup%>%filter(param=='main_output_folder_prefix')%>%pull(value)
main_output_folder<-glue('./output/{main_output_prefix}.bam_coverage_analysis.{Sys.Date()}')
if (!dir.exists(main_output_folder)){dir.create(main_output_folder)}
bed_analysis_output_folder<-glue('{main_output_folder}/bed_analysis')
bam_analysis_output_folder<-glue('{main_output_folder}/bam_analysis')
if (!dir.exists(bed_analysis_output_folder) & analysis_type%in%c('bed','both')){dir.create(bed_analysis_output_folder)}
if (!dir.exists(bam_analysis_output_folder) & analysis_type%in%c('bam','both')){dir.create(bam_analysis_output_folder)}
# copy analysis setup to the ouput folder
system(glue('cp {analysis_setup_file} {main_output_folder}'))
```

```{r bed_files_intersect}
bed_coverage_metrics<-NULL
if (analysis_type%in%c('bed','both')){
  if (class(bam_target_names)!='character'){
    message(glue('WARNING: Analysis type marked as {analysis_type} but no bam target files provided. will not check the theoretical target coverage'))
    stop()
  }
  for (i in 1:nrow(bam_targets_table)){
    bam_target_name<-bam_target_names[i]
    bam_target<-bam_targets[i]
    for (j in 1:nrow(target_regions_table)){
      target_region<-target_regions[j]
      target_region_name<-target_regions_names[j]
      message(glue('Analyzing how {bam_target_name} covers {target_region_name}'))
      bed_name<-glue('{target_region_name}_vs_{bam_target_name}')
      bed_coverage_output<-run_bedtools_coverage_by_bed(bed_a = target_region,
                                                        bed_b = bam_target,
                                                        output_folder=bed_analysis_output_folder,
                                                        bed_name = bed_name)
      # parse the coverage output and collect all the target regions that do not have full coverage 
      bed_coverage<-readr::read_delim(bed_coverage_output,delim='\t',col_names = c('chr','start','end','mean_cov'))
      bed_coverage<-bed_coverage%>%
        mutate(target_length=end-start,
               bases_not_covered=target_length*(1-mean_cov))
      bed_coverage$total_target_length<-sum(bed_coverage$target_length)
      bed_coverage<-bed_coverage%>%mutate(rate_out_of_total_target=bases_not_covered/total_target_length)
      target_bed_regions_not_covered_in_bam_target<-bed_coverage%>%filter(mean_cov<1)
      
      bed_coverage_metrics<-bed_coverage_metrics%>%
        bind_rows(data.frame(target_bed_regions_not_covered_in_bam_target,target_name=target_region_name,bam_target=bam_target_name))
    }
  }
  write.table(bed_coverage_metrics,file=glue('{bed_analysis_output_folder}/coverage_by_target_beds.{Sys.Date()}.txt'),sep='\t',row.names = F)
  cov_summary<-
    bed_coverage_metrics%>%group_by(target_name,bam_target)%>%
    summarize(bases_not_covered=sum(bases_not_covered),
              rate_not_covered=sum(rate_out_of_total_target))
  write.table(cov_summary,file=glue('{bed_analysis_output_folder}/coverage_by_target_beds_summary.{Sys.Date()}.txt'),sep='\t',row.names = F)
}


```

```{r calculate_bam_coverage_with_bedtools}
if (analysis_type%in%c('bed','both')){
  bam_analysis_df<-data.frame(input_bams_df,genome_file='',coverage_output='',target_region='')
  for (i in 1:nrow(target_regions_table)){
    target_region<-target_regions_table%>%slice(i)%>%pull(file)
    target_region_name<-target_regions_table%>%slice(i)%>%pull(name)
    for (group_name in unique(input_bams_df$group_name)){
      message(glue('Analyzing {group_name}..'))
      #if (group_name!='idt'){next}
      group_output_folder<-glue('{bam_analysis_output_folder}/{group_name}')
      if (!dir.exists(group_output_folder)){dir.create(group_output_folder)}
      for (sample_name in input_bams_df[input_bams_df$group_name==group_name,'sample_name']){
        message(glue('Analyzing {sample_name}..'))
        sample_output_folder<-glue('{group_output_folder}/{sample_name}')
        if (!dir.exists(sample_output_folder)){dir.create(sample_output_folder)}
        input_bam<-input_bams_df[input_bams_df$group_name==group_name & input_bams_df$sample_name==sample_name,'sample_path']
        genome_file<-prepare_genome_file(input_bam,sample_output_folder)
        sorted_target_bed<-sort_input_bed_file(target_region,sample_output_folder,genome_file)
        # Run the bedtools coverage analysis #### 
        coverage_ouput<-run_bedtools_coverage_by_bam(input_bam,genome_file,sorted_target_bed,sample_output_folder,remove_duplicates=remove_duplicates)
        bam_analysis_df<-bam_analysis_df%>%rows_update(data.frame(target_region=target_region_name,
                                                                  group_name=group_name,
                                                                  sample_name=sample_name,
                                                                  genome_file=genome_file,
                                                                  coverage_output=coverage_ouput),
                                                       by = c("group_name","sample_name"))
      }
    }
  }
  bam_analysis_table<-glue('{bam_analysis_output_folder}/bam_analysis_table.csv')
  write.table(bam_analysis_df,file=bam_analysis_table,sep='\t',row.names = F)
}
```

go over the target file and output problematic regions for each input file

```{r analyze_coverage_by_target}
output_folder_to_parse<-bam_analysis_output_folder

bam_analysis_def_table<-readr::read_delim(glue('{output_folder_to_parse}/bam_analysis_table.csv'))
threshs<-c(1,10,20,50)
all_cov_output<-NULL
per_target_cov_output<-NULL
for (i in 1:nrow(bam_analysis_def_table)){
  a_cov_output<-NULL
  pt_cov_output<-NULL
  group_name<-bam_analysis_def_table$group_name[i]
  sample_name<-bam_analysis_def_table$sample_name[i]
  target_region<-bam_analysis_def_table$target_region[i]
  bam_cov_file<-bam_analysis_def_table$coverage_output[i]
  bam_cov<-readr::read_delim(bam_cov_file,delim = '\t',col_names = c('chr','start','end','depth','num_bases','total_bases_in_target','percent_of_target'))
  all_cov<-bam_cov%>%filter(chr=='all')
  colnames(all_cov)<-c('all','depth','num_bases','total_bases_in_target','percent_of_entire_target','X1','X2')
  
  bam_cov<-bam_cov%>%filter(chr!='all')
  # per-target-mean-coverage
  pt_mean_cov<-bam_cov%>%group_by(chr,start,end)%>%summarize(min_cov=min(depth),max_cov=max(depth),mean_cov=sum(depth*percent_of_target))
  a_mean_cov<-all_cov%>%summarize(min_cov=min(depth),max_cov=max(depth),mean_cov=sum(depth*percent_of_entire_target))
  
  for (thresh in threshs){
    
    bases_col_name<-glue('bases_below_{thresh}x')
    pct_col_name<-glue('pct_below_{thresh}x')
    # All target analysis
    all_below_thresh<-all_cov%>%
      filter(depth<thresh)%>%
      group_by(all)%>%
      summarize(!!sym(bases_col_name):=sum(num_bases),
                !!sym(pct_col_name):=sum(percent_of_entire_target))
  
    # Per target analysis
    below_thresh<-bam_cov%>%
      filter(depth<thresh)%>%
      group_by(chr,start,end,total_bases_in_target)%>%
      summarize(!!sym(bases_col_name):=sum(num_bases),
                !!sym(pct_col_name):=sum(percent_of_target))
    if(is.null(pt_cov_output)){
      pt_cov_output<-below_thresh 
      a_cov_output<-all_below_thresh
    }else{
      pt_cov_output<-pt_cov_output%>%right_join(below_thresh)%>% replace(is.na(.), 0)
      a_cov_output<-a_cov_output%>%right_join(all_below_thresh)%>%replace(is.na(.), 0)
    }
  }
  a_cov_output<-a_cov_output%>%mutate(sample=sample_name,group=group_name,target_region=target_region)%>%bind_cols(a_mean_cov)
  pt_cov_output<-pt_cov_output%>%mutate(sample=sample_name,group=group_name,target_region=target_region)%>%left_join(pt_mean_cov)
per_target_cov_output<-per_target_cov_output%>%bind_rows(pt_cov_output)
all_cov_output<-all_cov_output%>%bind_rows(a_cov_output)
}
all_cov_output
write.table(all_cov_output,file=glue('{output_folder_to_parse}/target_coverage_analysis.all.{Sys.Date()}.csv'))
write.table(per_target_cov_output,file=glue('{output_folder_to_parse}/target_coverage_analysis.per_target.{Sys.Date()}.csv'))

xlsx::write.xlsx(as.data.frame(all_cov_output),sheetName = 'All targets coverage',row.names = FALSE,file=glue('{output_folder_to_parse}/target_coverage_analysis.{Sys.Date()}.xls'))
xlsx::write.xlsx(as.data.frame(per_target_cov_output),sheetName = 'Per target coverage',row.names = FALSE,file=glue('{output_folder_to_parse}/target_coverage_analysis.{Sys.Date()}.xls'),append = TRUE)

```
