---
title: "Gene coverage analysis"
author: "Genomics Center at Clalit"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: united
    toc: true
    toc_float: false
---
```{r initial_setup,include=F}
analysis_or_report='report'
picard_analysis_output_folder<-'/media/SSD/Bioinformatics/Projects/gene_coverage_analysis_2022/output/picard_coverage_analysis_2023-08-20/'
```


```{r render_report,eval=F,echo=F}
library(rmarkdown)
render('/media/SSD/Bioinformatics/Projects/gene_coverage_analysis_2022/src/picard_coverage_analysis_report.Rmd', 
       output_dir = picard_analysis_output_folder,
       output_file = glue('picard_coverage_report.{Sys.Date()}.html'),quiet = F)
```

```{r setup, include=FALSE,warning=F,message=F}

project_dir<-'/media/SSD/Bioinformatics/Projects/gene_coverage_analysis_2022/'
knitr::opts_knit$set(root.dir='/media/SSD/Bioinformatics/Projects/gene_coverage_analysis_2022/')
knitr::opts_chunk$set(echo = F)
library(ProjectTemplate)
setwd(project_dir)
load.project()

```

```{r load_data,message=F}
message(glue('Parsing PICARD output: {picard_analysis_output_folder}'))
coverage_df<-parse_picard_output_folder(picard_analysis_output_folder)%>%
  mutate(group_name=factor(group_name),
         sample_name=factor(sample_name),
         BAIT_SET=factor(BAIT_SET))
```

## Per-Sample Coverage Analysis Results Table

```{r ps_coverage_analysis_table,warning=F,message=F}
# DT::datatable(coverage_df,
#               # datatable definitions
#               options=list(scrollX=F,pageLength=10,dom = 'Bfrtip', buttons = I('colvis')),
#               rownames= FALSE,
#               filter = list(position = 'top', clear = FALSE),
#               extensions = 'Buttons')

DT:::datatable(
    coverage_df,
    filter = "top",
    rownames= F,
    extensions = list("Buttons" = NULL,
                      "FixedColumns" = NULL),
    options = list(
                scrollX=T,
                #columnDefs = list(list(targets = c(3:ncol(coverage_df)-1), visible = FALSE)),
                fixedColumns = list(leftColumns = 2),
                dom = 'BRrltpi',
                autoWidth=TRUE,
                lengthMenu = list(c(10, 50, -1), c('10', '50', 'All')),
                buttons = list('colvis')
              )
    )


```

## Aggregated Coverage Analysis Results Table

```{r ag_coverage_analysis_table,warning=F,message=F}
agg_coverage_df<-coverage_df%>%group_by(group_name,BAIT_SET)%>%summarize(across(where(is.numeric),mean))
DT::datatable(agg_coverage_df,
              filter = "top",
              rownames= F,
              # datatable definitions
              extensions = list("Buttons" = NULL,
                                "FixedColumns" = NULL),
              options = list(
                          scrollX=T,
                          #columnDefs = list(list(targets = c(3:ncol(coverage_df)-1), visible = FALSE)),
                          fixedColumns = list(leftColumns = 2),
                          dom = 'BRrltpi',
                          autoWidth=TRUE,
                          lengthMenu = list(c(10, 50, -1), c('10', '50', 'All')),
                          buttons = list('colvis')
                        )
              )
```

## Percent covered above X 

```{r percent_above_x,warning=F,message=F,fig.width=15,fig.height=15}
plot_percent_covered_above_x(coverage_df,threshs = c(1,10,20,50))
```

## Base metrics plot

```{r base_metrics,warning=F,message=F,fig.width=15,fig.height=15}
plot_basic_seq_and_map_metrics(coverage_df)
```

## Metrics Plots {.tabset}

```{r, results = 'asis',fig.width=12,fig.height=8,eval=T,include=T}
numeric_cols<-coverage_df%>%select(where(is.numeric))%>%colnames()
excluded_cols<-c('BAIT_TERRITORY','BAIT_DESIGN_EFFICIENCY','HS_LIBRARY_SIZE','GENOME_SIZE','TARGET_TERRITORY')
for (metric in setdiff(numeric_cols,excluded_cols)){
  cat("### ", metric, "\n")
  p<-plot_per_target_metric(coverage_df,metric)
  #cat(knitr::knit_print(print(p)))
  knitr::knit_print(p)
  cat("\n\n")
}
```
