---
title: "gene coverage analysis"
author: "Clalit Genomics"
date: "14/01/2022"
output: 
  html_document:
    theme: united
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
analysis_setup_file<-'./config/analysis_setup/twist_local_vs_external.analysis_setup.csv'
project_dir<-'/media/SSD/Bioinformatics/Projects/gene_coverage_analysis_2022/'
knitr::opts_knit$set(root.dir=project_dir)
knitr::opts_chunk$set(echo = F)
library(ProjectTemplate)
setwd(project_dir)
load.project()
```

```{r prep_analysis}
# prep output
main_output_prefix<-analysis_setup%>%filter(param=='main_output_folder_prefix')%>%pull(value)
main_output_folder<-glue('./output/{main_output_prefix}.picard_coverage_analysis.{Sys.Date()}')
if (!dir.exists(main_output_folder)){dir.create(main_output_folder)}

# copy analysis setup to the ouput folder
system(glue('cp {analysis_setup_file} {main_output_folder}'))

```

```{r bam_coverage_analysis_picard}
# fix target regions - remove all chromosomes found in the region and not in the dict
target_regions<-fix_target_regions(target_regions,reference_dict,main_output_folder)

# Verify that target regions are intervals
target_intervals<-c()
target_intervals_output_folder<-glue('{main_output_folder}/target_intervals')
dir.create(target_intervals_output_folder)
for (target_region in target_regions){
  message(target_region)
  target_interval<-run_BedToIntervalList(input_bed = target_region,
                                         output_folder = target_intervals_output_folder,
                                         #output_folder = './data/target_intervals',
                                         reference_dict = reference_dict)
  target_intervals<-c(target_intervals,target_interval)
}

target_intervals_df<-data.frame(target_intervals,per_target_analysis_req)
bam_analysis_df<-data.frame(input_bams_df)
coverage_df<-NULL

for (group_name in unique(input_bams_df$group_name)){
  message(glue('Analyzing {group_name}..'))
  #if (group_name!='idt'){next}
  group_output_folder<-glue('{main_output_folder}/{group_name}')
  if (!dir.exists(group_output_folder)){dir.create(group_output_folder)}
  for (sample_name in input_bams_df[input_bams_df$group_name==group_name,'sample_name']){
    message(glue('Analyzing {sample_name}..'))
    sample_output_folder<-glue('{group_output_folder}/{sample_name}')
    if (!dir.exists(sample_output_folder)){dir.create(sample_output_folder)}

    input_bam<-input_bams_df[input_bams_df$group_name==group_name & input_bams_df$sample_name==sample_name,'sample_path']
    for (target_interval in target_intervals){
      target_interval_name<-stringr::str_replace(basename(target_interval),'.intervals.*','')
      analyze_per_target<-target_intervals_df[target_intervals_df$target_interval==target_interval,'per_target_analysis_req']
      per_target_output_file<-NA
      if (analyze_per_target=='Y'){
        per_target_output_file<-glue('{sample_output_folder}/{sample_name}_{target_interval_name}_picard_per_target_cov_metrics.txt')
      }
      picard_cov_output_file<-
        glue('{sample_output_folder}/{sample_name}_{target_interval_name}_picard_cov_metrics.txt')
      if (!file.exists(picard_cov_output_file)){
        run_CollectHsMetrics(input_bam,picard_cov_output_file,reference_fasta,target_interval,target_interval,per_target_output_file)
      }else{
        message(glue('Found coverage analysis output file: {picard_cov_output_file}, will not run it again'))
      }
      picard_cov_output<-parse_picard_output(picard_cov_output_file)
      coverage_df<-coverage_df%>%plyr::rbind.fill(data.frame(group_name=group_name,
                                                                sample_name=sample_name,
                                                                picard_cov_output[['picard_coverage_metrics']]))
    }
  }
}

coverage_file_name<-glue('{main_output_folder}/picard_coverage_analysis_results_{Sys.Date()}.tsv')
write.table(coverage_df,file=coverage_file_name,sep='\t',row.names = F)
```

```{r generate_picard_coverage_report}
library(rmarkdown)
library(knitr)
render('/media/SSD/Bioinformatics/Projects/gene_coverage_analysis_2022/src/picard_coverage_analysis_report.Rmd', 
       output_dir = main_output_folder,
       output_file = glue('{main_output_folder}/{main_output_prefix}.picard_coverage.{Sys.Date()}.html'),
       quiet = F,
       params=list(main_output_folder=main_output_folder,
                   main_output_prefix=main_output_prefix))
```



```{r analyze_results}
#main_output_folder<-glue('./output/enrichment_kit_comp_picard_coverage_analysis_2022-06-19')
picard_results_file<-grep('picard_coverage_analysis_results',list.files(main_output_folder),value=T)
picard_results<-readr::read_delim(glue('{main_output_folder}/{picard_results_file}'))
rois<-picard_results%>%pull(BAIT_SET)%>%unique()

#rois<-c('refseq_hg19_curated_cds_20220414','agilent_XT_HS_v8_hg19','hg19_Twist_Exome_2','xgen-exome-research-panel-v2-targets-hg19')

cols_to_summarize<-c('MEAN_TARGET_COVERAGE',
                     'PCT_EXC_DUPE',
                     'PCT_TARGET_BASES_20X',
                     'ZERO_CVG_TARGETS_PCT')

for (roi in rois){
  g<-
    picard_results%>%filter(BAIT_SET==roi)%>%
    select(group_name,cols_to_summarize)%>%
    rename('Mean coverage'=MEAN_TARGET_COVERAGE,
           'Percent duplicates'=PCT_EXC_DUPE,
           'Percent >20x'=PCT_TARGET_BASES_20X,
           'Percent not covered'=ZERO_CVG_TARGETS_PCT)%>%
    select(group_name,`Percent >20x`)%>%
    pivot_longer(-group_name)%>%
    ggplot(aes(x=group_name,fill=group_name,y=value))+
    geom_boxplot(color='black',alpha=0.75)+geom_jitter(size=1,alpha=0.5,width = 0.1)+
    facet_wrap(.~name,scales='free_y',nrow = 2)+
    ggsci::scale_fill_nejm()+ggsci::scale_color_nejm()+
    theme_minimal()+labs(fill=NULL,title=roi)+
    theme(legend.position = 'top',axis.text.x = element_blank())
  g
  ggsave(g,file=glue('{main_output_folder}/{roi}_coverage_plots.jpg'),
       device = 'jpg',
       width = 10,height = 10,dpi=150,bg='white')
}

```
